classifier.py

```diff
@@ -3130,59 +3090,436 @@ def target_age(cls, identifier, name):
+            target_age_min, target_age_max = target_age_max, target_age_min
+        return target_age_min, target_age_max
+
+    def genres(self, fiction, cutoff=0.15):
+        """Consolidate genres and apply a low-pass filter."""
+        # Remove any genres whose fiction status is inconsistent with the
+        # (independently determined) fiction status of the book.
+        #
+        # It doesn't matter if a book is classified as 'science
+        # fiction' 100 times; if we know it's nonfiction, it can't be
+        # science fiction. (It's probably a history of science fiction
+        # or something.)
+        genres = dict(self.genre_weights)
+        for genre in list(genres.keys()):
+            if genre.default_fiction != fiction:
+                del genres[genre]
+
+        # Consolidate parent genres into their heaviest subgenre.
+        genres = self.consolidate_genre_weights(genres)
+        total_weight = float(sum(genres.values()))
+
+        # Strip out the stragglers.
+        for g, score in list(genres.items()):
+            affinity = score / total_weight
+            if affinity < cutoff:
+                total_weight -= score
+                del genres[g]
+        return genres
+
+    def weigh_genre(self, genre_data, weight):
+        """A helper method that ensure we always use database Genre
+        objects, not GenreData objects, when weighting genres.
+        """
+        from model import Genre
+        genre, ignore = Genre.lookup(self._db, genre_data.name)
+        self.genre_weights[genre] += weight
+
+    @property
+    def most_reliable_target_age_subset(self):
+        """Not all target age data is created equal. This method isolates the
+        most reliable subset of a set of classifications.
+        
+        For example, if we have an Overdrive classification saying
+        that the book is a picture book (target age: 0-3), and we also
+        have a bunch of tags saying that the book is for ages 2-5 and
+        0-2 and 1-3 and 12-13, we will use the (reliable) Overdrive
+        classification and ignore the (unreliable) tags altogether,
+        rather than try to average everything out.
+        
+        But if there is no Overdrive classification, that set of tags
+        will be the most reliable target age subset, and we'll
+        just do the best we can.
+        """
+        highest_quality_score = None
+        reliable_classifications = []
+        for c in self.target_age_relevant_classifications:
+            score = c.quality_as_indicator_of_target_age
+            if not score:
+                continue
+            if (not highest_quality_score or score > highest_quality_score):
+                # If we gather a bunch of data, then discover a more reliable
+                # type of data, we need to start all over.
+                highest_quality_score = score
+                reliable_classifications = []
+            if score >= highest_quality_score:
```

Comment:
I think it would be clearer to say score == highest_quality_score since it should never be greater when it gets here.